securedrop/journalist.py

```diff
@@ -575,18 +575,38 @@ def download_single_submission(sid, fn):
 @app.route('/reply', methods=('POST',))
 @login_required
 def reply():
+    msg = request.form['msg']
+
+    # Reject empty replies
+    if not msg:
+        flash("You cannot send an empty reply!", "error")
+        return redirect(url_for('col', sid=g.sid))
+
     g.source.interaction_count += 1
     filename = "{0}-{1}-reply.gpg".format(g.source.interaction_count,
                                           g.source.journalist_filename)
-    crypto_util.encrypt(request.form['msg'],
+    crypto_util.encrypt(msg,
                         [crypto_util.getkey(g.sid), config.JOURNALIST_KEY],
                         output=store.path(g.sid, filename))
     reply = Reply(g.user, g.source, filename)
-    db_session.add(reply)
-    db_session.commit()
 
-    flash("Thanks! Your reply has been stored.", "notification")
-    return redirect(url_for('col', sid=g.sid))
+    try:
+        db_session.add(g.source)
```

Comment:
I realized right after posting this that what I was saying wasn't consistent, but needed to investigate some things in order to properly correct myself. If it's true that query results are implicitly tracked, then there's really no significance in explicitly calling `db_session.add(g.source)`. That is, it should neither solve nor create problems; everything should behave the same under the same conditions regardless of the presence of that line. Note, I have omitted it now because it's redundant, not for other reasons discussed above.

In the source is deleted scenario, no exception will be raised--regardless of when the deletion of the source happens between when the query for the source object in `journalist.get_source()` succeeds and when `db_session.commit()` is called in `journalist.reply()`--until `db_session.commit()` is called. In this interim period the state of the session is detached from the state of the database. If the source is deleted a `StaleDataError` will be raised upon that call to `db_session.commit()`.

I've changed the error handling to take this into account. Now, when this exception is encountered, the database is rolled back, and then a query is made against the database to confirm that this was the cause of the exception (namely, our source has been deleted). If so, an appropriate message is flashed and the journalists is returned to the index view.